_deploy/r/gnoswap/gns/gns.gno

```diff
@@ -12,43 +12,79 @@ import (
 	"gno.land/r/demo/grc20reg"
 	"gno.land/r/demo/users"
 
-	"gno.land/r/gnoswap/v1/common"
 	"gno.land/r/gnoswap/v1/consts"
 )
 
-const MAXIMUM_SUPPLY = uint64(1_000_000_000_000_000) // 1B
+const (
+	MAXIMUM_SUPPLY      = uint64(1_000_000_000_000_000)
+	INITIAL_MINT_AMOUNT = uint64(100_000_000_000_000)
+	MAX_EMISSION_AMOUNT = uint64(900_000_000_000_000) // MAXIMUM_SUPPLY - INITIAL_MINT_AMOUNT
+)
+
+var (
+	lastMintedHeight = std.GetHeight()
+)
 
 var (
-	lastMintedHeight int64
-	amountToEmission uint64
+	leftEmissionAmount   = MAX_EMISSION_AMOUNT
+	mintedEmissionAmount = uint64(0)
 )
 
 var (
 	Token, privateLedger = grc20.NewToken("Gnoswap", "GNS", 6)
 	UserTeller           = Token.CallerTeller()
-	owner                = ownable.NewWithAddress("g17290cwvmrapvp869xfnhhawa8sm9edpufzat7d") // ADMIN
+	owner                = ownable.NewWithAddress(consts.ADMIN)
 )
 
 func init() {
-	privateLedger.Mint(owner.Owner(), 100_000_000_000_000) // 100_000_000 GNS ≈ 0.1B
+	privateLedger.Mint(owner.Owner(), INITIAL_MINT_AMOUNT)
 	getter := func() *grc20.Token { return Token }
 	grc20reg.Register(getter, "")
+}
+
+// MintedEmissionAmount returns the amount of GNS that has been minted by the emission contract.
+// It does not include initial minted amount.
+func MintedEmissionAmount() uint64 {
+	return TotalSupply() - INITIAL_MINT_AMOUNT
+}
 
-	amountToEmission = MAXIMUM_SUPPLY - uint64(100_000_000_000_000)
+func MintGns(address pusers.AddressOrName) uint64 {
+	lastMintedHeight := GetLastMintedHeight()
+	currentHeight := std.GetHeight()
 
-	lastMintedHeight = std.GetHeight()
+	// skip minting process if gns for current block is already minted
+	if skipIfSameHeight(lastMintedHeight, currentHeight) {
+		return 0
+	}
+
+	assertShouldNotBeHalted()
+	assertCallerIsEmission()
+
+	// calculate gns amount to mint, and the mint to the target address
+	amountToMint := calculateAmountToMint(lastMintedHeight+1, currentHeight)
+	err := privateLedger.Mint(users.Resolve(address), amountToMint)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	// update
+	setLastMintedHeight(currentHeight)
+	setMintedEmissionAmount(GetMintedEmissionAmount() + amountToMint)
+	setLeftEmissionAmount(GetLeftEmissionAmount() - amountToMint)
+
+	return amountToMint
 }
 
-func GetAmountToEmission() uint64 { return amountToEmission }
+func Burn(from pusers.AddressOrName, amount uint64) {
```

Comment:
In case we ever use `Burn`, can you make sure to add extra tests that:
- `leftEmissionAmount` will not be affected by the burnt amount (e.g, the burnt amount was already in circulating supply, so it shouldn't affect the amount to be minted)
- `mintedEmissionAmount` will be affected by the burnt amount (e.g, the burnt amount should be deducted in `mintedEmissionAmount` so that `INITIAL_MINT_AMOUNT` + `mintedEmissionAmount` = TotalSupply (Circulating Supply, not MAX).

Reply:
c143862711ce2f118c1b12389fcda96767176d6f

Reply:
testcase screenshot for above commit.
![image](https://github.com/user-attachments/assets/d1e78731-c939-4bc3-ac9d-ec9c95bf93a5)